src/Microsoft.AspNetCore.Http.Abstractions/Routing/EndpointMetadataCollection.cs

```diff
@@ -0,0 +1,178 @@
+    /// of arbitrary types. The metadata items are stored as an ordered collection with
+    /// items arranged in ascending order of precedence.
+    /// </remarks>
+    public sealed class EndpointMetadataCollection : IReadOnlyList<object>
+    {
+        /// <summary>
+        /// An empty <see cref="EndpointMetadataCollection"/>.
+        /// </summary>
+        public static readonly EndpointMetadataCollection Empty = new EndpointMetadataCollection(Array.Empty<object>());
+
+        private readonly object[] _items;
+
+        /// <summary>
+        /// Creates a new <see cref="EndpointMetadataCollection"/>.
+        /// </summary>
+        /// <param name="items">The metadata items.</param>
+        public EndpointMetadataCollection(IEnumerable<object> items)
+        {
+            if (items == null)
+            {
+                throw new ArgumentNullException(nameof(items));
+            }
+
+            _items = items.ToArray();
+        }
+
+        /// <summary>
+        /// Creates a new <see cref="EndpointMetadataCollection"/>.
+        /// </summary>
+        /// <param name="items">The metadata items.</param>
+        public EndpointMetadataCollection(params object[] items)
+            : this((IEnumerable<object>)items)
+        {
+        }
+
+        /// <summary>
+        /// Gets the item at <paramref name="index"/>.
+        /// </summary>
+        /// <param name="index">The index of the item to retrieve.</param>
+        /// <returns>The item at <paramref name="index"/>.</returns>
+        public object this[int index] => _items[index];
+
+        /// <summary>
+        /// Gets the count of metadata items.
+        /// </summary>
+        public int Count => _items.Length;
+
+        /// <summary>
+        /// Gets the most significant metadata item of type <typeparamref name="T"/>.
+        /// </summary>
+        /// <typeparam name="T">The type of metadata to retrieve.</typeparam>
+        /// <returns>
+        /// The most significant metadata of type <typeparamref name="T"/> or <c>null</c>.
+        /// </returns>
+        public T GetMetadata<T>() where T : class
+        {
+            for (var i = _items.Length - 1; i >= 0; i--)
+            {
+                var item = _items[i] as T;
+                if (item != null)
+                {
+                    return item;
+                }
+            }
+
+            return default;
+        }
+
+        /// <summary>
+        /// Gets the metadata items of type <typeparamref name="T"/> in ascending
+        /// order of precedence.
+        /// </summary>
+        /// <typeparam name="T">The type of metadata.</typeparam>
+        /// <returns>A sequence of metadata items of <typeparamref name="T"/>.</returns>
+        public IEnumerable<T> GetOrderedMetadata<T>() where T : class
+        {
+            for (var i = 0; i < _items.Length; i++)
+            {
+                var item = _items[i] as T;
+                if (item != null)
```

Comment:
`if (_items[i] is T item)`

Reply:
https://github.com/aspnet/Routing/pull/708